src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -1312,37 +1314,195 @@ private ArrayBuilder<string> ExtractStringValuesFromAttributes(List<AttributeInf
+                BlobHandle valueBlob = GetCustomAttributeValueOrThrow(handle);
+                if (!valueBlob.IsNil)
+                {
+                    BlobReader reader = MetadataReader.GetBlobReader(valueBlob);
+
+                    if (reader.Length >= 4)
                     {
-                        return new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError: false);
+                        // check prolog
+                        if (reader.ReadInt16() == 1)
+                        {
+                            var (diagnosticId, urlFormat) = CrackObsoleteProperties(ref reader);
+                            return new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message: null, isError: false, diagnosticId, urlFormat);
+                        }
                     }
+                }
+            }
+            catch (BadImageFormatException)
+            { }
 
-                    return null;
+            return null;
+        }
 
-                case 2:
-                    // ObsoleteAttribute(string, bool)
-                    return TryExtractValueFromAttribute(attributeInfo.Handle, out var obsoleteData, s_attributeObsoleteDataExtractor) ?
-                        obsoleteData :
-                        null;
+        /// <summary>
+        /// Decodes attribute argument type from attribute blob (called FieldOrPropType in the spec).
+        /// </summary>
+        /// <exception cref="UnsupportedSignatureContent">If the encoded argument type is invalid.</exception>
+        /// <exception cref="BadImageFormatException">An exception from metadata reader.</exception>
+        private static (SerializationTypeCode typeCode, SerializationTypeCode elementTypeCode) DecodeCustomAttributeFieldOrPropTypeOrThrow(ref BlobReader argReader)
+        {
+            var typeCode = argReader.ReadSerializationTypeCode();
+
+            // Spec:
+            // The FieldOrPropType (typeCode) shall be exactly one of: ELEMENT_TYPE_BOOLEAN,
+            // ELEMENT_TYPE_CHAR, ELEMENT_TYPE_I1, ELEMENT_TYPE_U1, ELEMENT_TYPE_I2,
+            // ELEMENT_TYPE_U2, ELEMENT_TYPE_I4, ELEMENT_TYPE_U4, ELEMENT_TYPE_I8,
+            // ELEMENT_TYPE_U8, ELEMENT_TYPE_R4, ELEMENT_TYPE_R8, ELEMENT_TYPE_STRING.
+            // 
+            // A single-dimensional, zero-based array is specified as a single byte 0x1D followed
+            // by the FieldOrPropType of the element type. (See §II.23.1.16) An enum is
+            // specified as a single byte 0x55 followed by a SerString.
+            // 
+
+            // todo: this could be simplified for the "skip" case.
+            if (typeCode == SerializationTypeCode.SZArray)
+            {
+                var elementTypeCode = argReader.ReadSerializationTypeCode();
+                if (elementTypeCode == SerializationTypeCode.SZArray)
+                {
+                    // nested array not allowed:
+                    throw new UnsupportedSignatureContent();
```

Comment:
>throw new UnsupportedSignatureContent(); [](start = 20, length = 40)

Who is going to handle this exception? It doesn't look like this type was ever throwing an exception of this type. Especially not code dealing with attribute cracking. We should find a way to communicate failure in some other way. For example, we can return SerializationTypeCode.Invalid in this case.

Reply:
We will return a 'null' ObsoleteAttributeData instead of throwing UnsupportedSignatureContent or BadImageFormatException.